packages/backend-api/install/tests/ShopBundle/Smoke/BackendApiUpdateProductTest.php

```diff
@@ -13,30 +14,41 @@
  */
 class BackendApiUpdateProductTest extends OauthTestCase
 {
+    /**
+     * @var \Shopsys\FrameworkBundle\Component\Domain\Domain
+     */
+    private $domain;
+
+    protected function setUp()
+    {
+        parent::setUp();
+        $this->domain = $this->getContainer()->get(Domain::class);
+    }
+
     public function testUpdateProductCompletely(): void
     {
         $uuid = $this->createProduct();
 
+        $namesByLocale = [];
+        $shortDescriptionsByDomainId = [];
+        $longDescriptionsByDomainId = [];
+        foreach ($this->domain->getAll() as $domainConfig) {
+            $namesByLocale[$domainConfig->getLocale()] = 'Changed name';
+            $shortDescriptionsByDomainId[$domainConfig->getId()] = 'Changed short description';
+            $longDescriptionsByDomainId[$domainConfig->getId()] = 'Changed long description';
+        }
```

Comment:
It cannot be extracted in separate method because of translations. They can be dumped only if they are written in code.

Reply:
That is right, but you are not using translations in this class. Maybe it is a bug?

Even though you can still separate thhese chunks in a method and pass already translated name to it.

Reply:
We do not need translations here. We just need to have the values set somehow for all locales. The exactly same loop is used twice in the test and then once again with different values. So how can I extract it? Are you suggesting a method like that?
```
getNamesIndexedByLocaleAndShortDescriptionsAndDescriptionsIndexedByDomainId($string $nameVaue, string $shortDescriptionValue, $descriptionValue)
``` 
And what should it return, a value object containing all the three arrays? 

I think we are overthinking this, it is just a test.

Reply:
In this class there are only 3 occurences, but it is used many times as a variation elsewhere too. This is just the one I pointed to.

I would suggest creating method 
```
getProductLocalizedDataForAllDomains(
    ['name' => ..., 'shortDesc...' => ..., 'longDesc...' => ...],
    $domainConfigs
)
```
returning an array
```
[
  'name' => [
    'cs' => ....,
    'en' => ...,
  ],
  'shortDesc' => [
    1 => ....,
    2 => ....,
  ],
  ...
]
```

then use `array_merge()` for creating a final array

Reply:
I tried your suggestion in e5fcca615 and I still do not think it is better :disappointed: IMHO it is not worth more effort

Reply:
It can be nicely done in `createProduct()` which could be using the same logic as in `BackendApiCreateProductTest::getValidProduct()`.

In `BackendApiCreateProductTest::testValidationError()` you can obtain a valid product and then just define breaaking changes (array with the same structure of keys, but wrong values) and replace it via [`array_replace_recursive()`](https://www.php.net/manual/en/function.array-replace-recursive.php).

`$invalidProduct = array_replace_recursive($validProduct, $invalidValues);`

What do you thing?

Reply:
IMHO it worsens the readability of the tests, @DavidOstrozlik, please decide it for us :smile:

Reply:
It depends... Personally it would be much cleaner to see what is actually different from valid product.

Yours feelings will decide.

Reply:
I agree with @vitek-rostislav, that it adds bit too much confusion and worsens the readability and it doesn't seem to be worth it for what it does..

Reply:
@s3tezsky requested to use only one definition of product and change it when it is necessary

Reply:
I have refactored this a little in d15f5c0f2142bfa759927e3e69896adc971aef0a

Reply:
and 203e4aead6063261abb67c2cc25745a3cf76a35a